src/core/solver.py

```
@@ -327,6 +380,86 @@ def add_cohesion_penalties(self, groupers: dict[str, set[int]]) -> None:
+
+        for i, p in enumerate(self.participants):
+            # Sort scores and tags for stable, order-insensitive hashing
+            sorted_scores = tuple(sorted(p.scores.items()))
+            identity = (
+                sorted_scores,
+                tuple(sorted(TagProcessor.get_tags(p.groupers))),
+                tuple(sorted(TagProcessor.get_tags(p.separators))),
+            )
+            identity_map.setdefault(identity, []).append(i)
+
+        for _, indices in identity_map.items():
```

Ruff (0.15.12)

[warning] When using only the values of a dict use the `values()` method

Replace `.items()` with `.values()`

(PERF102)